packages/web-components/fast-foundation/src/tooltip/tooltip.ts

```diff
@@ -0,0 +1,352 @@
+import { attr, DOM, FASTElement, observable } from "@microsoft/fast-element";
+import { AnchoredRegion, AxisPositioningMode, AxisScalingMode } from "../anchored-region";
+import { keyCodeEscape } from "@microsoft/fast-web-utilities";
+
+export type TooltipPosition = "top" | "right" | "bottom" | "left";
+
+const hiddenRegionStyle: string = `
+     display: none;
+     pointer-events: none;
+`;
+
+const visibleRegionStyle: string = `
+     pointer-events: none;
+`;
+
+export class Tooltip extends FASTElement {
+    @attr({ mode: "boolean" })
+    public visible: boolean;
+    private visibleChanged(): void {
+        if ((this as any).$fastController.isConnected) {
+            this.updateTooltipVisibility();
+            this.updateLayout();
+        }
+    }
+
+    @attr
+    public anchor: string = "";
+    private anchorChanged(): void {
+        if ((this as any).$fastController.isConnected) {
+            this.updateLayout();
+        }
+    }
+
+    @attr
+    public delay: number = 300;
+
+    @attr
+    public position: TooltipPosition | null = null;
+    private positionChanged(): void {
+        if ((this as any).$fastController.isConnected) {
+            this.updateLayout();
+        }
+    }
+
+    @observable
+    public anchorElement: HTMLElement | null = null;
+    private anchorElementChanged(oldValue: HTMLElement | null): void {
+        if ((this as any).$fastController.isConnected) {
+            if (oldValue !== null && oldValue !== undefined) {
+                oldValue.removeEventListener("mouseover", this.handleAnchorMouseOver);
+                oldValue.removeEventListener("mouseout", this.handleAnchorMouseOut);
+            }
+
+            if (this.anchorElement !== null && this.anchorElement !== undefined) {
+                this.anchorElement.addEventListener(
+                    "mouseover",
+                    this.handleAnchorMouseOver,
+                    { passive: true }
+                );
+                this.anchorElement.addEventListener(
+                    "mouseout",
+                    this.handleAnchorMouseOut,
+                    { passive: true }
+                );
+
+                const anchorId: string = this.anchorElement.id;
+
+                document.querySelectorAll(":hover").forEach(element => {
```

Comment:
Can't rely on being able to find it in the document. Might need to use `this.getRootNode()`.

Reply:
Even when wrapped in a "$fastController.isConnected" check?

Reply:
No, because if your component is inside of another component's shadow dom, a selector executed against the document will not be able to fin it. You need to get the root node (which might be the document...but could be another shadow root) and then run the query selector on that.

Reply:
So whenever we want to call something on document from code in a web component we should recursively call getRootNode until the return type is HTMLDocument?

Reply:
You don't need to recursively call it. Calling it once will get you the nearest shadow root that your element is in, or the document if it's not in a shadow DOM.

Reply:
Maybe I misunderstood the code...are you trying to find all :hover in the same "scope" as the current element? If so, then you would use getRootNode()

Reply:
No, I'm actually trying to figure out if the anchor is being hovered at that point in time. It could be anywhere in the document.

Reply:
In the document or in the DOM scope that the tooltip is being used within?

Reply:
There can be N doms in a web page. The document is one but so is each shadow DOM. The tooltip could exist directly in the document or also inside of a shadow dom. My assumption is that the tooltip is used in the same dom as its anchor, in which case you would need to use getRootNode(). If the anchor can be anywhere...then we probably need to change the design of this component or at least introduce the restriction that the tooltip and the anchor need to exist in the same DOM.

Reply:
Ok, so I can query the anchor element itself to determine hover, which is actually way better than querying the document.  But there are other places where we query the document for elements by id to get the anchor and viewport.  I think that's ok as that's the most basic use case and we present more advanced authors who may be working in a shadow dom with the ability to pass in those elements directly via JS.

Reply:
I think that sounds fine.